lib/marin/src/marin/processing/classification/dedupe.py

```diff
@@ -397,8 +406,89 @@ def _load_batches(file_path: str, columns: list[str] | None = None, **parquet_kw
             yield from pa_json.read_json(f).to_batches()
 
 
+def _load_dupe_map_shard(shards: list[str]) -> dict[str, dict[str, str]]:
+    shard_dup_map = {}
+
+    def add_to_dup_map(record: dict):
+        shard_dup_map[record["hash"]] = {"canonical": record["canonical"]}
+
+    with log_time(f"Load duplicate map from {len(shards)} shards"):
+        create_backend("threadpool").execute(
+            Dataset.from_list(shards)
+            .flat_map(lambda p: load_parquet(p, columns=["hash", "canonical"]))
+            # NOTE: would be nice if Zephyr could optimize the predicate pushdown for Parquet
+            .filter(lambda record: record["hash"] is not None)
+            .map(add_to_dup_map)
+        )
+
+    return shard_dup_map
+
+
+@dataclass
+class Counters:
+    total: int = 0
+    unique: int = 0
+    unique_dups: int = 0
+
+    def __add__(self, other: "Counters") -> "Counters":
+        assert isinstance(other, Counters)
+
+        return Counters(
+            total=self.total + other.total,
+            unique=self.unique + other.unique,
+            unique_dups=self.unique_dups + other.unique_dups,
+        )
+
+
+def _compute_dedup_stats(shards: list[str]) -> Counters:
+    with log_time(f"Compute deduplication stats from {len(shards)} shards"):
+        return create_backend("threadpool").execute(
+            Dataset.from_list(shards)
+            .flat_map(lambda p: load_parquet(p, columns=["cnt"]))
+            .map(lambda c: Counters(total=c["cnt"], unique=int(c["cnt"] == 1), unique_dups=int(c["cnt"] > 1)))
+            .reduce(partial(sum, start=Counters()))
+        )[0]
+
+
+class DupeReduceResult(typing.TypedDict):
+    hash: str | None
+    cnt: int
+    canonical: str | None
+
+
+def _count_reduce(key: str, items: Iterator[pa.StructScalar], *, canonical_id: str) -> DupeReduceResult:
+    head = next(items)
+    doc_cnt = sum(map(lambda _: 1, items)) + 1
+    if doc_cnt == 1:
+        return {
+            "hash": None,
+            "cnt": 1,
+            "canonical": None,
+        }
+
+    return {
+        "hash": key,
+        "cnt": doc_cnt,
+        "canonical": head[canonical_id],
+    }
+
+
+def _find_base_path(input_path: str | list[str], input_files: list[str]) -> str:
+    # Determine base path for rebasing
+    base_path = input_path[0] if isinstance(input_path, list) else input_path
+    if base_path in input_files:
+        # NOTE: if the base_path is in the input_files, means it's a specific file, so rebase to its directory
+        base_path = os.path.dirname(base_path)
+    return base_path
+
+
 def _run_deduplication(config: DedupeConfig):
-    input_files = collect_input_files(config.input_path)
+    import dupekit
```

Comment:
when we have a bunch of these, i often try to do:

try:
  from Y import X
except ImportError:
  X: Any = None

but i acknowledge this can annoy the type-checker

Reply:
Hopefully this as well goes away with https://github.com/marin-community/marin/pull/2129#discussion_r2593957905 🤞